test/integration/shoots/operations/shoot_op_test.go

```diff
@@ -0,0 +1,128 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package operations_test
+
+import (
+	"context"
+	"flag"
+	"strings"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/labels"
+
+	"github.com/gardener/gardener/pkg/apis/garden/v1beta1"
+	"github.com/gardener/gardener/pkg/logger"
+	. "github.com/gardener/gardener/test/integration/framework"
+	. "github.com/gardener/gardener/test/integration/shoots"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	kubeconfig        = flag.String("kubeconfig", "", "the path to the kubeconfig  of Garden the cluster that will be used for integration tests")
+	shootTestYamlPath = flag.String("shootpath", "", "the path to the shoot yaml that will be used for testing")
+	testShootsPrefix  = flag.String("prefix", "", "prefix to use for test shoots")
+	shootNamespace    = flag.String("shootNamespace", "", "shootNamespace is used for creation and deletion of shoots")
+	logLevel          = flag.String("verbose", "", "verbosity level, when set, logging level will be DEBUG")
+	cleanup           = flag.Bool("cleanup", false, "cleans up all created test shoots after test")
+)
+
+const (
+	WaitForShootCreationTimeout = 3600 * time.Second
+	WaitForShootDeletionTimeout = 3600 * time.Second
+	WaitForCreateDeleteTimeout  = 7200 * time.Second
+)
+
+var _ = Describe("Shoot creation/deletion testing", func() {
+	var (
+		shootGardenerTest         *ShootGardenerTest
+		shoot                     *v1beta1.Shoot
+		shootOperationsTestLogger *logrus.Logger
+	)
+
+	BeforeSuite(func() {
+		if len(*kubeconfig) == 0 {
+			Fail("The path to the Garden cluster kubeconfig is required, please specify it and try again")
+		}
+
+		// parse shoot yaml into shoot object and generate random test names for shoots
+		_, shootObject, err := CreateShootTestArtifacts(*shootTestYamlPath, *testShootsPrefix)
+		Expect(err).To(BeNil())
+
+		shoot = shootObject
+		shootOperationsTestLogger = logger.AddWriter(logger.NewLogger(*logLevel), GinkgoWriter)
+
+		shootGardenerTest, err = NewShootGardenerTest(*kubeconfig, shoot, shootOperationsTestLogger)
+		Expect(err).To(BeNil())
+	})
+
+	AfterSuite(func() {
+		if *cleanup {
+			if len(*shootNamespace) == 0 {
+				Fail(ErrShootsNamespaceWasNotSpecified.Error())
+			}
+
+			shootGardenerTest.Logger.Info("Cleaning up all test shoots!")
+			shoots, err := shootGardenerTest.ShootLister.Shoots(*shootNamespace).List(labels.Everything())
```

Comment:
This is a bit hard - if I execute tests on the same cluster concurrently, I'd start deleting Shoots of other clusters - you have to think of some better criteria than this.

Reply:
that's the intention, it is for garbage collection. If garbage collection is not preferable you can simply disable it and build your own clean-up logic after the tests. I.e., this is a luxury that is not required by the test.

Reply:
Yes but one test interfering with another is _architecturally wrong_: The automated tests will all run with garbage collection enabled, if one test then suddenly kills shoots of another test this is a bug. And no, I'd not say that garbage collection is a luxury, it _should_ be the default and can only be disabled for debugging purposes, otherwise manual intervention and resources leaking is just a few test runs ahead.